src/AutocompleteBadges.js

```diff
@@ -0,0 +1,145 @@
+'use strict';
+
+import templates from './AutocompleteBadges.soy.js';
+import Autocomplete from 'metal-autocomplete';
+import Component from 'metal-component';
+import Soy from 'metal-soy';
+
+class AutocompleteBadges extends Component {
+
+	/**
+	 * @inheritDoc
+	 */
+	attached() {
+		this.inputElement = this.element.querySelector('.autocomplete-field');
+		this.listBadgesElement = this.element.querySelector('.autocomplete-badges--list');
+
+		this.autocomplete_ = new Autocomplete({
+			inputElement: this.inputElement,
+			data: this.getFilteredElements_.bind(this)
+		});		
+
+		this.autocomplete_.on('select', this.onListItemSelected_.bind(this));
+
+		this.element.style.width = this.inputElement.offsetWidth + 'px';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	rendered() {
+		if(this.listBadgesElement) {			
+			this.inputElement.style.paddingLeft = this.getFullWidthListBadges_() + 'px';	
+		}
+	}
+
+	/**
+	 * Recover a full width of basges list	
+	 * @return {Number} returns the width
+	 * @protected
+	 */
+	getFullWidthListBadges_() {
+		return this.listBadgesElement.offsetWidth + this.listBadgesElement.style.marginLeft +
+				this.listBadgesElement.style.marginRight + this.listBadgesElement.style.paddingLeft + 
+				this.listBadgesElement.style.paddingRight;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	disposed() {
+		if(this.autocomplete_) {
+			this.autocomplete_.dispose();
+		}
+	}
+
+	/**
+	 * Recover a filtered list the according with the query search
+	 * @param {String} query
+	 * @return {Array} returns the filtered list
+	 * @protected
+	 */
+	getFilteredElements_(query) {
+		return this.elements.filter(function(item) {
+			return query && item.toLowerCase().includes(query.toLowerCase());
+		}).sort();
+	}
+
+	/**
+	 * Remove element from the list and add in badges list	 
+	 * @param {!Element} item The list selected item.
+	 * @protected
+	 */
+	onListItemSelected_(item) {
+		let index,
```

Comment:
We don't require variables to be defined at the top of the function for metal. We actually prefer them to be defined next to the place that would use them. So, for both of these, they could be defined right before the call to `this.elements.splice`, like this:

```js
const index = this.elements.indexOf(item.text);
const element = this.elements[index];
```